### PR TITLE
Add spectral norm op

### PR DESCRIPTION
**Pull request by heavengate**
https://arxiv.org/pdf/1802.05957.pdf

### CODE REVIEW

File: paddle/fluid/operators/spectral_norm_op.cc [hunk 1]
```diff
@@ -0,0 +1,180 @@
+/* Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserve.
```
Comment (jerrywgz) [inline paddle/fluid/operators/spectral_norm_op.cc]: 2019

File: paddle/fluid/operators/spectral_norm_op.cu [hunk 2]
```diff
@@ -0,0 +1,22 @@
+/* Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserve.
```
Comment (jerrywgz) [inline paddle/fluid/operators/spectral_norm_op.cu]: 2019

File: paddle/fluid/operators/spectral_norm_op.cc [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+             "The input weight tensor of spectral_norm operator, "
+             "This can be a 2-D, 3-D, 4-D, 5-D tensor which is the"
+             "weights of fc, conv1d, conv2d, conv3d layer.");
+    AddInput("U",
+             "The weight_u tensor of spectral_norm operator, "
+             "This can be a 1-D tensor in shape [H, 1],"
+             "H is the 1st dimentions of Weight after reshape"
+             "corresponding by Attr(dim).");
+    AddInput("V",
+             "The weight_u tensor of spectral_norm operator, "
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/spectral_norm_op.cc]: weight_u-》weight_v ?

File: paddle/fluid/operators/spectral_norm_op.cc [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+                   "epsilob for numerical stability in"
+                   "calculating norms")
+        .SetDefault(1e-12);
+
+    AddComment(R"DOC(
+          This layer calculate the spectral normalize value of weight of
+          fc, conv1d, conv2d, conv3d layers which should be 2-D, 3-D, 4-D, 5-D
+          tensor.
+
+          Spectral normalization stabilizes the training of critis in GANs
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/spectral_norm_op.cc]: critis  -》 critic?

File: paddle/fluid/operators/spectral_norm_op.cc [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+ public:
+  void Make() override {
+    AddInput("Weight",
+             "The input weight tensor of spectral_norm operator, "
+             "This can be a 2-D, 3-D, 4-D, 5-D tensor which is the"
+             "weights of fc, conv1d, conv2d, conv3d layer.");
+    AddInput("U",
+             "The weight_u tensor of spectral_norm operator, "
+             "This can be a 1-D tensor in shape [H, 1],"
+             "H is the 1st dimentions of Weight after reshape"
```
Comment (jerrywgz) [inline paddle/fluid/operators/spectral_norm_op.cc:86]: mention the shape after reshape is better

File: paddle/fluid/operators/spectral_norm_op.h [hunk 6]
```diff
@@ -0,0 +1,271 @@
+/* Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserve.
```
Comment (jerrywgz) [inline paddle/fluid/operators/spectral_norm_op.h]: 2019

File: paddle/fluid/operators/spectral_norm_op.cc [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+                   "calculating norms")
+        .SetDefault(1e-12);
+
+    AddComment(R"DOC(
+          This layer calculate the spectral normalize value of weight of
+          fc, conv1d, conv2d, conv3d layers which should be 2-D, 3-D, 4-D, 5-D
+          tensor.
+
+          Spectral normalization stabilizes the training of critis in GANs
+          (Generative Adversarial Networks). This layers rescaling weight tensor
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/spectral_norm_op.cc]: layers-> layer

File: paddle/fluid/operators/spectral_norm_op.cc [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+          wiht spectral normalize value.
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/spectral_norm_op.cc]: wiht->with

File: paddle/fluid/operators/spectral_norm_op.h [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+  auto weight_t = EigenTensor<T, 2>::From(*weight);
+  auto u_t = EigenTensor<T, 2>::From(*u);
+  auto v_t = EigenTensor<T, 2>::From(*v);
+
+  const int h = weight->dims()[0];
+  const int w = weight->dims()[1];
+
+  for (int i = 0; i < power_iters; i++) {
+    blas.MatMul(*weight, true, *u, false, T(1), v, T(0));
+    auto v_t_norm =
```
Comment (jerrywgz) [inline paddle/fluid/operators/spectral_norm_op.h:78]: give some comment or formula here?

File: paddle/fluid/operators/spectral_norm_op.cc [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+             "This can be a 1-D tensor in shape [W, 1],"
+             "W is the 2nd dimentions of Weight after reshape"
+             "corresponding by Attr(dim).");
+    AddOutput("Out",
+              "The output weight tensor of spectral_norm operator, "
+              "This tensor is in same shape with Input(Weight).");
+
+    AddAttr<int>("dim",
+                 "dimension corresponding to number of outputs,"
+                 "default 0 for fc layer, and 1 for conv1d, conv2d, conv3d"
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/spectral_norm_op.cc]: 这句话好像不太明白

File: paddle/fluid/operators/spectral_norm_op.cc [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+                 "number of power iterations to calculate"
+                 "spectral norm, default is 1.")
+        .SetDefault(1);
+    AddAttr<float>("eps",
+                   "epsilob for numerical stability in"
+                   "calculating norms")
+        .SetDefault(1e-12);
+
+    AddComment(R"DOC(
+          This layer calculate the spectral normalize value of weight of
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/spectral_norm_op.cc]: calculates

File: python/paddle/fluid/tests/unittests/test_spectral_norm_op.py [hunk 12]
```diff
@@ -0,0 +1,122 @@
+#   Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (jerrywgz) [inline python/paddle/fluid/tests/unittests/test_spectral_norm_op.py]: 2019

File: paddle/fluid/operators/spectral_norm_op.cc (hunk 3 above)
Comment (shippingwang) [inline paddle/fluid/operators/spectral_norm_op.cc]: weight_v ?

File: paddle/fluid/operators/spectral_norm_op.cu (hunk 2 above)
Comment (wanghaoshuang) [inline paddle/fluid/operators/spectral_norm_op.cu]: 2018->2019

File: paddle/fluid/operators/spectral_norm_op.h [hunk 13]
```diff
@@ -0,0 +1,271 @@
...
+      for (int i = 0; i < rank; i++) {
+        if (i < dim) {
+          perm.push_back(i + 1);
+        } else if (i == dim) {
+          perm.push_back(0);
+        } else {
+          perm.push_back(i);
+        }
+      }
+      weight_grad->mutable_data<T>(dims, ctx.GetPlace());
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/spectral_norm_op.h:262]: 确定 weight_grad 不用初始化为0么？

File: paddle/fluid/operators/spectral_norm_op.h [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+      TransCompute<DeviceContext, T>(
+          rank, weight_grad_mat.Resize(framework::make_ddim(real_dims)),
+          weight_grad, perm, dev_ctx);
+    } else {
+      TensorCopySync(weight_grad_mat.Resize(dims), ctx.GetPlace(), weight_grad);
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/spectral_norm_op.h:267]: weight的gradients有可能来自多个op吧，所以当前op可以直接覆盖gradients么？还是改成累加？

File: paddle/fluid/operators/spectral_norm_op.h (hunk 9 above)
Comment (heavengate) [reply paddle/fluid/operators/spectral_norm_op.h:78]: Done

File: paddle/fluid/operators/spectral_norm_op.cc (hunk 5 above)
Comment (heavengate) [reply paddle/fluid/operators/spectral_norm_op.cc:86]: Done

File: paddle/fluid/operators/spectral_norm_op.cc (hunk 1 above)
Comment (heavengate) [reply paddle/fluid/operators/spectral_norm_op.cc]: Done

File: paddle/fluid/operators/spectral_norm_op.cc (hunk 3 above)
Comment (heavengate) [reply paddle/fluid/operators/spectral_norm_op.cc]: Done

Comment (heavengate) [reply paddle/fluid/operators/spectral_norm_op.cc]: Done

File: paddle/fluid/operators/spectral_norm_op.cc (hunk 10 above)
Comment (heavengate) [reply paddle/fluid/operators/spectral_norm_op.cc]: Done

File: paddle/fluid/operators/spectral_norm_op.cc (hunk 11 above)
Comment (heavengate) [reply paddle/fluid/operators/spectral_norm_op.cc]: Done

File: paddle/fluid/operators/spectral_norm_op.cc (hunk 4 above)
Comment (heavengate) [reply paddle/fluid/operators/spectral_norm_op.cc]: Done

File: paddle/fluid/operators/spectral_norm_op.cc (hunk 7 above)
Comment (heavengate) [reply paddle/fluid/operators/spectral_norm_op.cc]: Done

File: paddle/fluid/operators/spectral_norm_op.cc (hunk 8 above)
Comment (heavengate) [reply paddle/fluid/operators/spectral_norm_op.cc]: Done

File: paddle/fluid/operators/spectral_norm_op.cu (hunk 2 above)
Comment (heavengate) [reply paddle/fluid/operators/spectral_norm_op.cu]: Done

Comment (heavengate) [reply paddle/fluid/operators/spectral_norm_op.cu]: Done

File: paddle/fluid/operators/spectral_norm_op.h (hunk 6 above)
Comment (heavengate) [reply paddle/fluid/operators/spectral_norm_op.h]: Done

File: python/paddle/fluid/tests/unittests/test_spectral_norm_op.py (hunk 12 above)
Comment (heavengate) [reply python/paddle/fluid/tests/unittests/test_spectral_norm_op.py]: Done

**Comment by heavengate** [comment]
@haowang101779990 please check, thanks

File: python/paddle/fluid/layers/nn.py [hunk 15]
```diff
@@ -3346,6 +3347,94 @@ def group_norm(input,
     return helper.append_activation(group_norm_out)
 
 
+@templatedoc()
+def spectral_norm(weight, dim=0, power_iters=1, eps=1e-12, name=None):
+    """
+    **Spectral Normalization Layer**
+
+    This layer calculate the spectral normalize value of weight parameters of
```
Comment (haowang101779990) [inline python/paddle/fluid/layers/nn.py]:
```suggestion
    This layer calculates the spectral normalization value of weight parameters of
```

File: python/paddle/fluid/layers/nn.py [hunk 16]
```diff
@@ -3346,6 +3347,94 @@ def group_norm(input,
...
+    fc, conv1d, conv2d, conv3d layers which should be 2-D, 3-D, 4-D, 5-D
+    Parameters. Calculations are showed as followings.
```
Comment (haowang101779990) [inline python/paddle/fluid/layers/nn.py]:
```suggestion
    Parameters. Calculations are showed as follows.
```

File: python/paddle/fluid/layers/nn.py [hunk 17]
```diff
@@ -3346,6 +3347,94 @@ def group_norm(input,
...
+    Refer to `Spectral Normalization <https://arxiv.org/abs/1802.05957>`_ .
+
+    Args:
+        weight(${weight_type}): ${weight_comment}
+        dim(${dim_type}): ${dim_comment}
+        eps(${eps_type}): ${eps_comment}
+        name (str): The name of this layer. It is optional.
+
+    Returns:
+        Variable: A tensor variable of weight after spetral normalization.
```
Comment (haowang101779990) [inline python/paddle/fluid/layers/nn.py]:
```suggestion
        Variable: A tensor variable of weight parameters after spectral normalization.
```

File: python/paddle/fluid/layers/nn.py [hunk 18]
```diff
@@ -3346,6 +3347,94 @@ def group_norm(input,
...
+        While attr:`power_iters` is a positive interger, do following
+        iteration calculations with u and v for attr:`power_iters`
+        round.
+            \mathbf{v} = \mathbf{W}^{T} \mathbf{u}
+            \mathbf{v} = \frac{\mathbf{v}}{\|\mathbf{v}\|_2}
+            \mathbf{u} = \mathbf{W}^{T} \mathbf{v}
+            \mathbf{u} = \frac{\mathbf{u}}{\|\mathbf{u}\|_2}
+
+        Step 3:
+        Calculate \sigma{W} and scale weight values.
```
Comment (haowang101779990) [inline python/paddle/fluid/layers/nn.py]:
```suggestion
        Calculate \sigma{W} and normalize weight values.
```
'scale' may cause confusion. I think "normalize" is closer to our context

File: python/paddle/fluid/layers/nn.py [hunk 19]
```diff
@@ -3346,6 +3347,94 @@ def group_norm(input,
...
+    This layer calculate the spectral normalize value of weight parameters of
+    fc, conv1d, conv2d, conv3d layers which should be 2-D, 3-D, 4-D, 5-D
+    Parameters. Calculations are showed as followings.
+
+    .. code-block:: text
+
+        Step 1:
+        Generate vector u in shape of [h], and v in shape of [w].
+        While h is the attr:`dim`th dimension of the input weights,
+        and w is the product result of remain dimensions.
```
Comment (haowang101779990) [inline python/paddle/fluid/layers/nn.py]:
```suggestion
        and w is the product result of remaining dimensions.
```

File: paddle/fluid/operators/spectral_norm_op.cc [hunk 20]
```diff
@@ -0,0 +1,188 @@
...
+                 "weight of fc layer, and should be set as 1 if"
+                 "Input(Weight) is the weight of conv layer,"
+                 "default is 0.")
+        .SetDefault(0);
+    AddAttr<int>("power_iters",
+                 "number of power iterations to calculate"
+                 "spectral norm, default is 1.")
+        .SetDefault(1);
+    AddAttr<float>("eps",
+                   "epsilob for numerical stability in"
```
Comment (haowang101779990) [inline paddle/fluid/operators/spectral_norm_op.cc]:
```suggestion
                   "epsilon for numerical stability in"
```

**Review by haowang101779990 — COMMENTED** [review]
Here is my API effect review.

File: paddle/fluid/operators/spectral_norm_op.cc (hunk 20 above)
Comment (heavengate) [reply paddle/fluid/operators/spectral_norm_op.cc]: Done

File: python/paddle/fluid/layers/nn.py (hunk 15 above)
Comment (heavengate) [reply python/paddle/fluid/layers/nn.py]: Done

File: python/paddle/fluid/layers/nn.py (hunk 16 above)
Comment (heavengate) [reply python/paddle/fluid/layers/nn.py]: Done

File: python/paddle/fluid/layers/nn.py (hunk 19 above)
Comment (heavengate) [reply python/paddle/fluid/layers/nn.py]: Done

File: python/paddle/fluid/layers/nn.py (hunk 18 above)
Comment (heavengate) [reply python/paddle/fluid/layers/nn.py]: Done

File: python/paddle/fluid/layers/nn.py (hunk 17 above)
Comment (heavengate) [reply python/paddle/fluid/layers/nn.py]: Done

**Review by heavengate — COMMENTED** [review]
@haowang101779990 please check

File: paddle/fluid/operators/spectral_norm_op.cc [hunk 21]
```diff
@@ -0,0 +1,197 @@
...
+             "corresponding by Attr(dim). As for Attr(dim) = 1 "
+             "in conv2d layer with weight shape [M, C, K1, K2] "
+             "Weight will be reshape to [C, M*K1*K2], V will "
+             "be in shape [M*K1*K2, 1].");
+    AddOutput("Out",
+              "The output weight tensor of spectral_norm operator, "
+              "This tensor is in same shape with Input(Weight).");
+
+    AddAttr<int>("dim",
+                 "The index of dimention which should be permute "
```
Comment (haowang101779990) [inline paddle/fluid/operators/spectral_norm_op.cc]:
```suggestion
                 "The index of dimension which should be permuted "
```

File: paddle/fluid/operators/spectral_norm_op.cc [hunk 22]
```diff
@@ -0,0 +1,197 @@
...
+                 "to the first before reshape Input(Weight) to "
```
Comment (haowang101779990) [inline paddle/fluid/operators/spectral_norm_op.cc]:
```suggestion
                 "to the first before reshaping Input(Weight) to "
```

Comment (heavengate) [reply paddle/fluid/operators/spectral_norm_op.cc]: Done

File: paddle/fluid/operators/spectral_norm_op.cc (hunk 21 above)
Comment (heavengate) [reply paddle/fluid/operators/spectral_norm_op.cc]: Done

**Review by shanyi15 — DISMISSED** [review]
format and content of comment LGTM

**Review by wanghaoshuang — DISMISSED** [review]
LGTM.

**Review by shanyi15 — APPROVED** [review]

**Review by wanghaoshuang — APPROVED** [review]
LGTM.

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Review by panyx0718 — APPROVED** [review]

**Pull request merged**